calyx/src/passes/automatic_par.rs

```diff
@@ -0,0 +1,157 @@
+        .map(|en_comp| st.get_inst_index(en_comp))
+        // If any of the get_inst_index failed, return the error.
+        .collect::<Result<Vec<_>, _>>()?
+        .into_iter()
+        // any is shortcircuiting. It returnrs on the first true.
+        .any(|idx| {
+            // for every output port, check if any incoming/outgoing edges
+            // contain a component in `comps1`
+            st.graph[idx].out_ports().any(|port| {
+                st.connected_to(idx, port.to_string())
+                    .chain(st.connected_from(idx, port.to_string()))
+                    .any(|(node_data, _)| comps1.contains(node_data.get_name()))
+            })
+        }))
+}
+
+impl Visitor for AutomaticPar {
+    // use finish_seq so that we collapse things on the way
+    // back up the tree and potentially catch more cases
+    fn finish_seq(
+        &mut self,
+        seq: &ast::Seq,
+        comp: &mut Component,
+        _: &Context,
+    ) -> VisResult {
+        // get first enable statement and its index. this will act
+        // as the accumulator for the statement that we are collapsing
+        // things into
+        let (start, mut cmp_acc) = match seq.stmts.iter().enumerate().find_map(
+            |(i, stmt)| match stmt {
+                Control::Enable { data } => Some((i, data.clone())),
+                _ => None,
+            },
+        ) {
+            Some((s, c)) => (s, c),
+            None => return Ok(Action::Continue),
+        };
+
+        // vec of new control statements including all elements up to the first
+        // enable that we find
+        let mut new_stmts: Vec<ast::Control> = seq.stmts[..start].to_vec();
+
+        // start interation from the start+1 because the start is in `cmp_stmt`.
+        // This loop will keep trying to merge adjacent sequences as long as
+        // there are no conflicts. When it detects a conflict, it attempts
+        // to merge the sequence after the conflicting node.
+        for stmt in seq.stmts[start + 1..].iter() {
+            match stmt {
+                Control::Enable { data: enables2 } => {
+                    // for every component in the `enables2` we check if any
+                    // incoming/outgoing edge has an endpoint in `cmp_acc`
+                    if has_conflicts(
+                        &cmp_acc.comps,
+                        &enables2.comps,
+                        &comp.structure,
+                    )? {
+                        new_stmts.push(Control::enable(cmp_acc.comps.clone()));
+                        // there was a conflict, start, a new accumulator.
+                        cmp_acc = enables2.clone();
+                    } else {
+                        // If there is no conflict, update the current
+                        // component accumulator.
+                        cmp_acc.comps.append(&mut enables2.comps.clone());
+                    }
+                }
+                _ => new_stmts.push(stmt.clone()),
```

Comment:
I realized this is not quite correct. We also have to change the accumulator. Otherwise something like
```
(seq (enable A B)
      (seq ...)
      (enable C D))
```
would get collapsed to
```
(seq (seq ...)
      (enable A B C D))
```

Reply:
Ah, you’re right. Add a relevant test and feel free to fix if you get to it before me!